gh-119180: Add discussion of annotations to the 3.14 What's New (#124393)
python/cpython: Doc/whatsnew/3.14.rst

New Features
============

.. _whatsnew-314-pep649:

PEP 649: Deferred Evaluation of Annotations
-------------------------------------------

The :term:`annotations <annotation>` on functions, classes, and modules are no
longer evaluated eagerly. Instead, annotations are stored in special-purpose
:term:`annotate functions <annotate function>` and evaluated only when
necessary. This is specified in :pep:`649` and :pep:`749`.

The new :mod:`annotationlib` module provides tools for inspecting deferred
annotations. Annotations may be evaluated in the :attr:`~annotationlib.Format.VALUE`
format (which evaluates annotations to runtime values, similar to the behavior in
earlier Python versions), the :attr:`~annotationlib.Format.FORWARDREF` format
(which replaces undefined names with special markers), and the

> :attr:`~annotationlib.Format.SOURCE` format (which returns annotations as strings).

Reviewer: These will render as `VALUE`, `FORWARDREF`, `SOURCE`. Is that what we want, or would `Format.VALUE`, `Format.FORWARDREF` and `Format.SOURCE` be better?

Author: I did intend these to render as `VALUE` etc. Do you think adding `Format` is better?

Reviewer: I think I'd weakly prefer `Format.VALUE`, but it might make the markup quite verbose, and I don't have a strong opinion :-)

This example shows how these formats behave::

   from annotationlib import get_annotations, Format

   def func(arg: Undefined):
       pass

   print(get_annotations(func, format=Format.VALUE))  # raises NameError
   print(get_annotations(func, format=Format.FORWARDREF))  # {'arg': ForwardRef('Undefined')}

>    print(get_annotations(func, format=Format.SOURCE))  # {'arg': 'Undefined'}

Reviewer: If we used a doctest here, I feel like we could "show rather than tell", and this would also allow us to assert in CI that the example is not outdated:

```suggestion
This example shows how these formats behave:

.. doctest::

   >>> from annotationlib import get_annotations, Format
   >>> def func(arg: Undefined):
   ...     pass
   >>> get_annotations(func, format=Format.VALUE)
   Traceback (most recent call last):
     ...
   NameError: name 'Undefined' is not defined
   >>> get_annotations(func, format=Format.FORWARDREF)
   {'arg': ForwardRef('Undefined')}
   >>> get_annotations(func, format=Format.SOURCE)
   {'arg': 'Undefined'}
```

> Implications if you write annotations in your code
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Implications for writers of annotations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Author: I'm not sure about this, it sort of sounds like we're talking about code generation tools. I was struggling with how to communicate "people who just put annotations on their function and run a type checker".

Reviewer: Hmm, yeah. Maybe "Implications for annotated code"...?

Author: That's a good one, let's go with that.
```suggestion
Implications for annotated code
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

> If you write annotations in your code (for example, for use with a static type

Reviewer: I don't know why but "write annotations" sounds quite odd to me. Maybe "define annotations"?

```suggestion
If you define annotations in your code (for example, for use with a static type
```

checker), then this change probably does not affect you: you can keep

> writing annotations the same way you did in previous versions of Python.

Reviewer: ```suggestion
writing annotations the same way you did with previous versions of Python.
```

You will likely be able to remove quoted strings in annotations, which are frequently
used for forward references. Similarly, if you use ``from __future__ import annotations``
to avoid having to write strings in annotations, you may well be able to

> remove it. However, if you rely on third-party libraries that read annotations,

Reviewer: ```suggestion
remove that import. However, if you rely on third-party libraries that read annotations,
```

> those libraries may need changes to support unquoted annotations.

Reviewer: ```suggestion
those libraries may need changes to support unquoted annotations before they
work as expected.
```

> Implications if you read ``__annotations__``
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Implications for readers of ``__annotations__``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

If your code reads the ``__annotations__`` attribute on objects, you may want
to make changes in order to support code that relies on deferred evaluation of

> annotations. For example, you may want to use the :func:`annotationlib.get_annotations`

Reviewer: "you may want to use the `annotations.get_annotations` with the `FORWARDREF` format" isn't quite grammatical:

```suggestion
annotations. For example, you may want to use :func:`annotationlib.get_annotations`
```

with the :attr:`~annotationlib.Format.FORWARDREF` format, as the :mod:`dataclasses`
module now does.

Related changes
^^^^^^^^^^^^^^^

> The changes in Python 3.14 are designed to minimize changes to code that contains

Reviewer: ```suggestion
The changes in Python 3.14 are designed to rework how ``__annotations__``
works at runtime while minimizing breakage to code that contains
```

> annotations in source code and to code that reads ``__annotations__``. However,
> there are many changes to the supporting code, and if you rely on undocumented
> details of the annotation behavior or on private functions in the standard library,
> your code may not work in Python 3.14. To safeguard your code against future changes,
> use only the documented functionality of the :mod:`annotationlib` module.

Reviewer: I'd concentrate on the outcome to the user here (lots of ways in which the code might break) rather than the reason why ("many changes to the supporting code"). "Many changes to the supporting code" feels like it anyway makes me ask more questions rather than answering any I might have had :-)

```suggestion
annotations in source code and to code that reads ``__annotations__``. However,
if you rely on undocumented details of the annotation behavior or on private
functions in the standard library, there are many ways in which your code may
not work in Python 3.14. To safeguard your code against future changes,
use only the documented functionality of the :mod:`annotationlib` module.
```

``from __future__ import annotations``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In Python 3.7, :pep:`563` introduced the ``from __future__ import annotations``
directive, which turns all annotations into strings. This directive is now
considered deprecated and it is expected to be removed in a future version of Python.
However, this removal will not happen until after Python 3.13, the last version of

> Python without deferred evaluation of annotations, loses support.

Reviewer: ```suggestion
Python without deferred evaluation of annotations, reaches its end of life.
```

In Python 3.14, the behavior of code using ``from __future__ import annotations``

> does not change.

Reviewer: ```suggestion
is unchanged.
```
